vumi/middleware/session_length.py

```diff
@@ -0,0 +1,101 @@
+# -*- test-case-name: vumi.middleware.tests.test_session_length -*-
+
+from confmodel import Config
+from confmodel.fields import ConfigDict, ConfigInt, ConfigText
+
+from twisted.internet.defer import inlineCallbacks, returnValue
+from twisted.internet import reactor
+
+from vumi.message import TransportUserMessage
+from vumi.middleware.base import BaseMiddleware
+from vumi.persist.txredis_manager import TxRedisManager
+
+
+class SessionLengthMiddlewareConfig(Config):
+    """
+    Configuration class for the session length middleware.
+    """
+
+    redis = ConfigDict("Redis config", default={})
+    timeout = ConfigInt("Redis key timeout (secs)", default=600)
+    field_name = ConfigText(
+        "Field name in message helper_metadata", default="session")
+
+
+class SessionLengthMiddleware(BaseMiddleware):
+    """ Middleware for storing the session length in the message.
+
+    Stores the start timestamp of a session in the message payload in
+    message['helper_metadata'][field_name]['session_start'] for all messages
+    in the session, as well as the end timestamp if the session in
+    message['helper_metadata'][field_name]['session_end'] if the message
+    marks the end of the session.
+
+    Configuration options:
+
+    :param dict redis:
+        Redis configuration parameters.
+    :param int timeout:
+        Redis key timeout (in seconds). Defaults to 600.
+    :param str field_name:
+        The field name to use when storing the timestamps in the message
+        helper_metadata. Defaults to 'session'.
+    """
+    SESSION_NEW, SESSION_CLOSE = (
+        TransportUserMessage.SESSION_NEW, TransportUserMessage.SESSION_CLOSE)
+
+    @inlineCallbacks
+    def setup_middleware(self):
+        config = SessionLengthMiddlewareConfig(self.config)
```

Comment:
Could we use `static=True` in the constructor to only get the static config fields?